packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -166,299 +134,270 @@ class Hero extends StatefulWidget {
-  VoidCallback _disposeCallback;
-
-  @override
-  bool get alwaysAnimate => config.alwaysAnimate;
-
-  @override
-  _HeroManifest _takeChild(Animation<double> currentAnimation) {
-    assert(mounted);
-    final RenderBox renderObject = context.findRenderObject();
-    assert(renderObject != null);
-    assert(!renderObject.debugNeedsLayout);
-    assert(renderObject.hasSize);
-    if (_placeholderSize == null) {
-      // We are a "from" hero, about to depart on a quest.
-      // Remember our size so that we can leave a placeholder.
-      _placeholderSize = renderObject.size;
-    }
-    final Point heroTopLeft = renderObject.localToGlobal(Point.origin);
-    final Point heroBottomRight = renderObject.localToGlobal(renderObject.size.bottomRight(Point.origin));
-    final Rect heroArea = new Rect.fromLTRB(heroTopLeft.x, heroTopLeft.y, heroBottomRight.x, heroBottomRight.y);
-    _HeroManifest result = new _HeroManifest(
-      key: _key, // might be null, e.g. if the hero is returning to us
-      config: config,
-      sourceStates: new HashSet<_HeroState>.from(<_HeroState>[this]),
-      currentRect: heroArea,
-      currentTurns: config.turns.toDouble()
-    );
-    if (_key != null)
-      setState(() { _key = null; });
-    return result;
-  }
 
-  void _setChild(GlobalKey value) {
-    assert(_key == null);
-    assert(_placeholderSize != null);
-    assert(value != null);
+  void startFlight() {
     assert(mounted);
+    final RenderBox box = context.findRenderObject();
+    assert(box != null && box.hasSize);
     setState(() {
-      _key = value;
-      _placeholderSize = null;
+      _placeholderSize = box.size;
     });
   }
 
-  void _resetChild() {
-    if (mounted)
-      _setChild(new GlobalKey());
-  }
-
-  @override
-  void dispose() {
-    if (_disposeCallback != null)
-      _disposeCallback();
-    super.dispose();
+  void endFlight() {
+    if (mounted) {
+      setState(() {
+        _placeholderSize = null;
+      });
+    }
   }
 
   @override
   Widget build(BuildContext context) {
     if (_placeholderSize != null) {
-      assert(_key == null);
-      return new SizedBox(width: _placeholderSize.width, height: _placeholderSize.height);
+      return new SizedBox(
+        width: _placeholderSize.width,
+        height: _placeholderSize.height
+      );
     }
     return new KeyedSubtree(
       key: _key,
-      child: config.child
+      child: config.child,
     );
   }
 }
 
-
-class _HeroQuestState implements _HeroHandle {
-  _HeroQuestState({
-    this.tag,
-    this.key,
-    this.child,
-    this.sourceStates,
-    this.animationArea,
-    this.targetRect,
-    this.targetTurns,
-    this.targetState,
-    this.currentRect,
-    this.currentTurns
+// Everything known about a hero flight that's to be started or restarted.
+class _HeroFlightManifest {
+  _HeroFlightManifest({
+    this.type,
+    this.overlay,
+    this.navigatorRect,
+    this.fromRoute,
+    this.toRoute,
+    this.fromHero,
```

Comment:
I definitely use them less than for public classes, but this one seems like a good candidate.  :)